src/Text/Pandoc/Readers/Markdown.hs

```diff
@@ -1741,6 +1742,41 @@ source = do
 linkTitle :: MarkdownParser String
 linkTitle = quotedTitle '"' <|> quotedTitle '\''
 
+wikilinkstuff :: MarkdownParser (F Inlines, String)
+wikilinkstuff = do
+  string "[["
+  (_, raw) <- withRaw $ charsInBalancedBrackets 2
+  guard $ not $ null raw
+  let deunder c = if c == '_' then ' ' else c
+      deiki s = case elemIndices '/' s of
+                  [] -> map deunder s
+                  xs -> map deunder (drop ((last xs) + 1) s)
+  case elemIndex '|' raw of
+    Nothing -> do
+      let rawstr = ((reverse . (drop 2) . reverse) raw)
+          srcstr = rawstr
+          labstr = deiki rawstr
+      src <- parseFromString (many anyChar) srcstr
+      lab <- parseFromString (trimInlinesF . mconcat <$> many inline) labstr
+      return (lab, src)
+    Just idx -> do
```

Comment:
The `Nothing` and `Just` cases share a lot of code.  That's usually a sign that things can be abstracted.  One way to do this would be to write a local functions which takes `srcstr` and `labstr` as input and returns `(lab, src)`.

The computation for `rawstr` is the same in both cases, so this should be defined in the previous `let` block.